src/libmongoc/src/mongoc/mongoc-bulkwrite.c

```diff
@@ -0,0 +1,1833 @@
+                        _bulkwriteexception_set_error (ret.exc, &error);
+                        goto batch_fail;
+                     }
+                  }
+
+                  BSON_ASSERT (bson_in_range_size_t_signed (idx));
+                  // `models_idx` is the index of the model that produced this result.
+                  size_t models_idx = (size_t) idx + ops_doc_offset;
+                  if (ok == 0) {
+                     bson_iter_t result_iter;
+                     has_write_errors = true;
+
+                     // Parse `code`.
+                     int32_t code;
+                     if (!lookup_int32 (result, "code", &code, "result", ret.exc)) {
+                        goto batch_fail;
+                     }
+
+                     // Parse `errmsg`.
+                     const char *errmsg;
+                     if (!lookup_string (result, "errmsg", &errmsg, "result", ret.exc)) {
+                        goto batch_fail;
+                     }
+
+                     // Parse optional `errInfo`.
+                     bson_t errInfo = BSON_INITIALIZER;
+                     if (bson_iter_init_find (&result_iter, result, "errInfo")) {
+                        if (!_mongoc_iter_document_as_bson (&result_iter, &errInfo, &error)) {
+                           _bulkwriteexception_set_error (ret.exc, &error);
+                           goto batch_fail;
+                        }
+                     }
+
+                     // Store a copy of the write error.
+                     _bulkwriteexception_set_writeerror (ret.exc, code, errmsg, &errInfo, models_idx);
+                  } else {
+                     // This is a successful result of an individual operation.
+                     // Server only reports successful results of individual
+                     // operations when verbose results are requested
+                     // (`errorsOnly: false` is sent).
+
+                     modeldata_t *md = &_mongoc_array_index (&self->arrayof_modeldata, modeldata_t, models_idx);
+                     // Check if model is an update.
+                     switch (md->op) {
+                     case MODEL_OP_UPDATE: {
+                        bson_iter_t result_iter;
+                        // Parse `n`.
+                        int32_t n;
+                        if (!lookup_int32 (result, "n", &n, "result", ret.exc)) {
+                           goto batch_fail;
+                        }
+
+                        // Parse `nModified`.
+                        int32_t nModified;
+                        if (!lookup_int32 (result, "nModified", &nModified, "result", ret.exc)) {
+                           goto batch_fail;
+                        }
+
+                        // Check for an optional `upserted._id`.
+                        const bson_value_t *upserted_id = NULL;
+                        bson_iter_t id_iter;
+                        if (bson_iter_init_find (&result_iter, result, "upserted")) {
```

Comment:
I know you struck this from the spec tests, but would this successfully report a BSON null since it's wrapped in a `bson_value_t`?